DigitalLearningSolutions.Web/Controllers/RegisterController.cs

```diff
@@ -0,0 +1,164 @@
+﻿namespace DigitalLearningSolutions.Web.Controllers
+{
+    using System;
+    using DigitalLearningSolutions.Data.Services;
+    using DigitalLearningSolutions.Web.Extensions;
+    using DigitalLearningSolutions.Web.Models;
+    using DigitalLearningSolutions.Web.ServiceFilter;
+    using DigitalLearningSolutions.Web.ViewModels.Register;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.AspNetCore.Mvc;
+
+    public class RegisterController : Controller
+    {
+        private const string CookieName = "RegistrationData";
+        private readonly ICentresService centresService;
+        private readonly IJobGroupsService jobGroupsService;
+
+        public RegisterController(ICentresService centresService, IJobGroupsService jobGroupsService)
+        {
+            this.centresService = centresService;
+            this.jobGroupsService = jobGroupsService;
+        }
+
+        public IActionResult Index()
+        {
+            if (User.Identity.IsAuthenticated)
+            {
+                return RedirectToAction("Index", "Home");
+            }
+
+            var delegateRegistrationData = TempData.Get<DelegateRegistrationData>();
+
+            if (delegateRegistrationData == null || !Request.Cookies.ContainsKey(CookieName))
+            {
+                var id = Guid.NewGuid();
+                delegateRegistrationData = new DelegateRegistrationData { Id = id };
+
+                Response.Cookies.Append(
+                    CookieName,
+                    id.ToString(),
+                    new CookieOptions
+                    {
+                        Expires = DateTimeOffset.UtcNow.AddDays(30)
+                    });
+                TempData.Set(delegateRegistrationData);
+
+                return View();
+            }
+
+            return View(delegateRegistrationData.RegisterViewModel);
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<DelegateRegistrationData>))]
+        [HttpPost]
+        public IActionResult Index(RegisterViewModel model)
+        {
+            if (!ModelState.IsValid)
+            {
+                return View(model);
+            }
+
+            var data = TempData.Get<DelegateRegistrationData>();
+            data!.RegisterViewModel = model;
+            TempData.Set(data);
+
+            return RedirectToAction("LearnerInformation");
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<DelegateRegistrationData>))]
+        [HttpGet]
+        public IActionResult LearnerInformation()
+        {
+            var data = TempData.Get<DelegateRegistrationData>();
+            var viewModel = data!.LearnerInformationViewModel;
+            ViewBag.Centres = centresService.GetActiveCentres();

```

Comment:
Ah, one thing I forgot to ask - is there a reason we're setting these on ViewBag rather than on a view model where they'd be typed?

Reply:
The main reason is that I didn't want to store them in TempData, as cookies have a 4096 byte limit and at the moment TempData is just a collection of all the ViewModels. We could get around this in a few different ways but it feels simplest to me to just chuck them into ViewBag when they're needed.

Reply:
Oh yes, I wasn't implying to keep them in TempData, just in a typed model. If it's a lot more work to add them, happy to leave it! 👍